File: frontend/services/feed.service.ts
import { apiClient } from '../lib/api-client';

export type FeedCategory = 'house_rent' | 'buy' | 'sell' | 'service' | 'help';

export interface Location {
    city: string;
    area: string;
    address: string;
}

export interface FeedPost {
    id: string;
    user_id: string;
    user_name?: string;
    mess_id: string;
    mess_name?: string;
    category: FeedCategory;
    title: string;
    description: string;
    location: Location;
    contact_info: string;
    price?: number;
    status: 'active' | 'sold' | 'closed';
    created_at: string;
    updated_at: string;
}

export interface CreatePostRequest {
    category: FeedCategory;
    title: string;
    description: string;
    location: Location;
    contact_info: string;
    price?: number;
}

export const feedService = {
    async createPost(post: CreatePostRequest): Promise<FeedPost> {
        const { data } = await apiClient.post('/feed', post);
        return data.data;
    },

    async listPosts(filters?: { category?: string; city?: string; area?: string }): Promise<FeedPost[]> {
        const params = new URLSearchParams();
        if (filters?.category) params.append('category', filters.category);
        if (filters?.city) params.append('city', filters.city);
        if (filters?.area) params.append('area', filters.area);

        const { data } = await apiClient.get(`/feed?${params.toString()}`);
        return data.data;
    },

    async getPost(id: string): Promise<FeedPost> {
        const { data } = await apiClient.get(`/feed/${id}`);
        return data.data;
    },
};

File: backend/internal/core/services/feed_service.go
package services

import (
	"amar-dera/internal/core/domain"
	"amar-dera/pkg/utils"
	"context"
	"errors"
	"time"
)

type FeedService struct {
	repo     domain.FeedRepository
	messRepo domain.MessRepository
	userRepo domain.UserRepository
}

func NewFeedService(repo domain.FeedRepository, messRepo domain.MessRepository, userRepo domain.UserRepository) *FeedService {
	return &FeedService{
		repo:     repo,
		messRepo: messRepo,
		userRepo: userRepo,
	}
}

func (s *FeedService) CreatePost(ctx context.Context, post *domain.FeedPost) (*domain.FeedPost, error) {
	// Populate User and Mess details
	user, err := s.userRepo.GetByID(ctx, post.UserID)
	if err != nil || user == nil {
		return nil, errors.New("user not found")
	}
	post.UserName = user.Name

	// If messID is not provided (from handler/request), use user's current mess
	if post.MessID == "" {
		post.MessID = user.CurrentMessID
	}

	if post.MessID != "" {
		mess, err := s.messRepo.GetByID(ctx, post.MessID)
		if err == nil && mess != nil {
			post.MessName = mess.Name
		}
	}

	post.ID = utils.GenerateID("POST", 8)
	post.Status = "active"
	post.CreatedAt = time.Now()
	post.UpdatedAt = time.Now()

	if err := s.repo.Create(ctx, post); err != nil {
		return nil, err
	}

	return post, nil
}

func (s *FeedService) ListPosts(ctx context.Context, category, city, area string) ([]domain.FeedPost, error) {
	filter := make(map[string]interface{})
	if category != "" {
		filter["category"] = category
	}
	if city != "" {
		filter["location.city"] = city
	}
	if area != "" {
		filter["location.area"] = area
	}
	filter["status"] = "active"

	return s.repo.List(ctx, filter)
}

func (s *FeedService) GetPost(ctx context.Context, id string) (*domain.FeedPost, error) {
	return s.repo.GetByID(ctx, id)
}
